sound/soc/sof/utils.c

```diff
@@ -184,3 +184,63 @@ int snd_sof_create_page_table(struct device *dev,
 	return pages;
 }
 EXPORT_SYMBOL(snd_sof_create_page_table);
+
+int snd_sof_dsp_core_power_up(struct snd_sof_dev *sdev, unsigned int core_mask)
+{
+	unsigned long mask = core_mask;
+	int ret, ret1;
+	u32 i, j;
+
+	/* power up all requested cores in core_mask */
+	for_each_set_bit(i, &mask, SOF_MAX_DSP_NUM_CORES) {
+		ret = snd_sof_dsp_core_get(sdev, i);
+		if (ret < 0)
+			goto err;
+	}
+
+	return 0;
+err:
+	/* power down the cores that were powered up */
+	for_each_set_bit(j, &mask, SOF_MAX_DSP_NUM_CORES) {
+		if (j >= i)
+			break;
+
+		/* break and return the original error in case of error */
+		ret1 = snd_sof_dsp_core_put(sdev, j);
+		if (ret1 < 0)
+			break;
+	}
+
+	return ret;
+}
+EXPORT_SYMBOL(snd_sof_dsp_core_power_up);
+
+int snd_sof_dsp_core_power_down(struct snd_sof_dev *sdev, unsigned int core_mask)
+{
+	unsigned long mask = core_mask;
+	int ret, ret1;
+	u32 i, j;
+
+	/* power down all requested cores in core_mask */
+	for_each_set_bit(i, &mask, SOF_MAX_DSP_NUM_CORES) {
+		ret = snd_sof_dsp_core_put(sdev, i);
+		if (ret < 0)
+			goto err;
+	}
+
+	return 0;
+err:
+	/* power up the cores that were powered down */
+	for_each_set_bit(j, &mask, SOF_MAX_DSP_NUM_CORES) {
+		if (j >= i)
+			break;
+
+		/* break and return the original error in case of error */
+		ret1 = snd_sof_dsp_core_get(sdev, j);
+		if (ret1 < 0)
+			break;
+	}
+
+	return ret;
+}
+EXPORT_SYMBOL(snd_sof_dsp_core_power_down);
```

Comment:
I find this patch difficult to review, it introduces new functions and replaces them at the same time. Can we split in two patches, for add and replace?

Reply:
hmm let me think about it. I thought about it too but found it hard to split. The problem is new new function name is the same as the old one.

Reply:
well exactly, use a different name otherwise it's really tricky to follow